contrib/demos/xtree/xtree.man
.TH XTREE 1 "Release 5" "X Version 11"
.SH NAME
xtree - tree graphing filter
.SH SYNOPSIS
.B "xtree"
[-option ...]
.SH DESCRIPTION
The \fIxtree\fP program is a filter that generates tree graphs from 
outline-style representations of data provided on the standard input.
Pressing on nodes in the tree causes their name to be printed on the
standard output in one of several formats.
.SH OPTIONS
This program accepts all of the standard X Toolkit command line options
as well as the following:
.TP 8
.B \-separator \fIstring\fP
This option specifies the character or string prefix for each line.  The
number of separators at the beginning of the line determines the level in
the tree.  This is the same as setting the \fBseparator\fP resource.  The
default value is a tab.
.TP 8
.B \-vertical
This option indicates that the tree should be displayed vertically instead
of horizontally.  This is a shorthand for setting the \fBgravity\fP resource
of the \fITree\fP widget to \fBnorth\fP.
.TP 8
.B \-iformat \fIformat-name\fP
This option specifies the format in which tree data on the standard input is
encoded.  The \fIformat-name\fP may be either \fBnumber\fP or \fBseparator\fP.
This is the same as setting the \fBinputFormat\fP resource.  The default value
is \fBseparator\fP.
.TP 8
.B \-oformat \fIformat-name\fP
This option specifies the format in which the names of nodes that have been 
pressed should be printed on the standard output.  The \fIformat-name\fP may 
be: \fBnumber\fP, \fBseparator\fP, \fBname\fP, or \fBresource\fP.  This is
the same as setting the \fBoutputFormat\fP resource.  The default value 
is \fBname\fP.
.TP 8
.B \-nbase \fInumber\fP
This option specifies the level of the highest nodes in the tree for
\fBnumber\fP output when \fBnumber\fP was not used for input .  
It is the same as setting the \fBnumberBase\fP resource.  The default is 1,
and this value is overridden by any \fBnumber\fP input.
.SH RESOURCES
The following application resources may be set in addition to the individual
widget resources described in the Athena Widget Set documentation:
.TP 8
.B "separator (\fPclass\fB Separator)"
This resource specifies a string that is used as a prefix on each input line
to indicate the depth of that line in the tree.  The default is a tab.
.TP 8
.B "numberBase (\fPclass\fB NumberBase)"
This resource specifies the number index of top-most nodes on output.  If the
input data is in \fBnumber\fP format, the first entry of the input will
override this value.  The default is 1.
.TP 8
.B "inputFormat (\fPclass\fB Format)"
This resource specifies the format in which tree data on the standard input is
encoded.  The \fIformat-name\fP may be either \fBnumber\fP or \fBseparator\fP
(which is the default).
.TP 8
.B "outputFormat (\fPclass\fB Format)"
This resource specifies the format in which the names of nodes that have been
pressed should be printed on the standard output.  The \fIformat-name\fP may 
be either \fBnumber\fP, \fBseparator\fP, \fBname\fP (the default), or 
.BR resource .
.SH FORMATS
Input and output data may have the following formats:
.TP 8
.B separator
In this format, lines are indented using the \fBseparator\fP resource to
indicate their depth in the tree.  A child is indicated by indenting with
one separator more than the preceeding line; a sibling is 
indicated by indenting with the same number of separators as the preceeding
line.  This format may be used for both input and output.
.TP 8
.B number
In this format, lines are prefixed with their depth in the tree followed by
a colon and optional whitespace.  If this format is only used for output, then
the \fBnumberBase\fP resource is used for the level of the first entry.  This
format may be used for both input and output.
.TP 8
.B name
In this output-only format, only the text of the node is printed when the
command button for the node is pressed.
.TP 8
.B resource
In this output-only format, the text of all node from the root down to the
pressed node are printed, separated by a period.  This is most useful when
looking at trees of resources.
.SH "WIDGET HIERARCHY"
Resources may be specified for the following Athena widgets (class name):
.sp
.nf
.ta .5i 1.0i 1.5i 2.0i 2.5i 3.0i 3.5i 4.0i 4.5i 5.0i 5.5i 6.0i 6.5i 7.0i
xtree (XTree)
	pane (Paned)
		buttonbox (Box)
			quit (Command)
	form (Form)
		porthole (Porthole)
			tree (Tree)
				\fIentry-name\fP (Command)
		panner (Panner)
.fi
.sp
where \fIentry-name\fP is the contents of the label with all non-alphanumeric
characters removed.
.SH "SEE ALSO"
X(1), Athena Widget Set documentation
.SH COPYRIGHT
Copyright 1990, Massachusetts Institute of Technology.
.br
See \fIX(1)\fP for a full statement of rights and permissions.
.SH AUTHOR
Jim Fulton, MIT X Consortium


contrib/demos/xtree/xtree.c
/*
 * $XConsortium: xtree.c,v 1.7 91/02/17 17:14:08 dave Exp $
 *
 * Copyright 1990 Massachusetts Institute of Technology
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in advertising or
 * publicity pertaining to distribution of the software without specific,
 * written prior permission.  M.I.T. makes no representations about the
 * suitability of this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 *
 * M.I.T. DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL M.I.T.
 * BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
 * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN 
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Author:  Jim Fulton, MIT X Consortium
 */

#include "xtree.h"			/* get Node and externs */

char *ProgramName;			/* for error messages */
XtAppContext app_con;			/* our Application Context */
Widget toplevel;			/* master widget */
Node rootnode;				/* invisible root of tree */
AppResources GlobalResources;		/* global resource data */

#define Offset(field) XtOffsetOf(AppResources, field)

static XtResource resources[] = {
    { "separator", "Separator", XtRString, sizeof(char *),
	Offset(separator), XtRString, (XtPointer) "\t" },
    { "numberBase", "NumberBase", XtRInt, sizeof(int),
	Offset(number_origin), XtRImmediate, (XtPointer) 1 },
    { "inputFormat", "Format", XtRDataFormat, sizeof(Format),
	Offset(input_format), XtRImmediate, (XtPointer) FORMAT_SEPARATOR },
    { "outputFormat", "Format", XtRDataFormat, sizeof(Format),
	Offset(output_format), XtRImmediate, (XtPointer) FORMAT_NAME },
};

#undef offset

static void ActionQuit(), ActionCirculatePanner();
static XtActionsRec actions[] = {
    { "Quit", ActionQuit },
    { "CirculatePanner", ActionCirculatePanner },
};

static Atom wm_delete_window;

static char *fallbacks[] = {
    "*allowShellResize: true",
    "*Porthole.top: ChainTop",
    "*Porthole.left: ChainLeft",
    "*Porthole.bottom: ChainBottom",
    "*Porthole.right:  ChainRight",
    "*Porthole.resizable: on",
    "*Panner.top: ChainTop",
    "*Panner.left: ChainLeft",
    "*Panner.bottom: ChainTop",
    "*Panner.right:  ChainLeft",
    "*Panner.resizable: on",
    "*Porthole*Box.BorderWidth: 0",
    "*Porthole*Box.HSpace: 0",
    "*Porthole*Box.VSpace: 0",
    "*Paned*allowResize: true",
    "*buttonbox.quit.Translations: #override \\n <Btn1Down>,<Btn1Up>: Quit() unset()",
    "*panner.Translations: #augment \\n <Btn3Down>: CirculatePanner()",
    "*tree.Translations: #augment \\n <Btn3Down>: CirculatePanner()",
    (char *) NULL,
};

static XrmOptionDescRec options[] = {
    { "-separator", "Separator", XrmoptionSepArg, (XtPointer) NULL },
    { "-iformat", "inputFormat", XrmoptionSepArg, (XtPointer) NULL },
    { "-oformat", "outputFormat", XrmoptionSepArg, (XtPointer) NULL },
    { "-vertical", "*Tree.Gravity", XrmoptionNoArg, (XtPointer) "north" },
    { "-blw", "*Tree.LineWidth", XrmoptionSepArg, (XtPointer) NULL },
    { "-bbw", "*Tree*Command*BorderWidth", XrmoptionSepArg, (XtPointer) NULL },
    { "-nbase", "numberBase", XrmoptionSepArg, (XtPointer) NULL },
};

static char *help_message[] = {
"where options include:",
"    -separator string                i/o separator character",
"    -vertical                        display tree vertically",
"    -iformat format                  input format (num or sep)",
"    -oformat format                  output format (num, sep, name, res)",
"    -nbase number                    base (0 or 1) for number format",
"",
"formats (abbreviation):",
"    number (num)                     level: label",
"    separator (sep)                  char char label",
"    name                             label (for output only)",
"    resource (res)                   parent.child.subchild (for output only)",
"",
NULL
};

static void usage ()
{
    char **msg;
    fprintf (stderr, "usage:\n\t%s [-options ...]\n\n", ProgramName);
    for (msg = help_message; *msg; msg++) {
	fprintf (stderr, "%s\n", *msg);
    }
    exit (1);
}

main (argc, argv)
    int argc;
    char **argv;
{
    ProgramName = argv[0];

    toplevel = XtAppInitialize (&app_con, "XTree", options, XtNumber(options),
				(Cardinal *)&argc, argv, fallbacks,
				(ArgList) NULL, ZERO);
    if (argc != 1) usage ();

    XtAddConverter (XtRString, XtRDataFormat, cvt_data_format,
		    (XtConvertArgList) NULL, ZERO);
    XtAppAddActions (app_con, actions, XtNumber(actions));
    XtAppAddActions
	(XtWidgetToApplicationContext(toplevel), actions, XtNumber(actions));
    XtOverrideTranslations
	(toplevel, XtParseTranslationTable ("<Message>WM_PROTOCOLS: Quit()"));
    XtGetApplicationResources (toplevel, (caddr_t) &GlobalResources,
			       resources, XtNumber(resources), NULL, ZERO);

    create_ui ();

    read_data (stdin);

    build_tree (&rootnode, (Widget) NULL, FALSE);

    adjust_panner ();

    wm_delete_window = XInternAtom(XtDisplay(toplevel), "WM_DELETE_WINDOW",
				   False);
    (void) XSetWMProtocols (XtDisplay(toplevel), XtWindow(toplevel),
                            &wm_delete_window, 1);

    XtAppMainLoop (app_con);
}

/* ARGSUSED */
static void ActionQuit (w, event, params, num_params)
    Widget w;
    XEvent *event;
    String *params;
    Cardinal *num_params;
{
    exit (0);
}

/* ARGSUSED */
static void ActionCirculatePanner (w, event, params, num_params)
    Widget w;
    XEvent *event;
    String *params;
    Cardinal *num_params;
{
    circulate_panner_tree();
}
